src/Sema.zig

```diff
@@ -12404,6 +12404,8 @@ fn zirVarExtended(
 
     const lib_name: ?[]const u8 = if (small.has_lib_name) blk: {
         const lib_name = sema.code.nullTerminatedString(sema.code.extra[extra_index]);
+        sema.owner_decl.has_lib_name = true;
+        sema.owner_decl.zir_lib_name_index = @intCast(Zir.Inst.Index, extra_index);
```

Comment:
This is storing a ZIR index into a Decl. A ZIR index changes when source files are updated - even when unrelated changes are made earlier on in the file. However a Decl is a long-lived object that persists beyond multiple updates. So if you want to hang on to a ZIR index then you have to take it into account into the `Module.mapOldZirToNew` mechanism, which tries to figure out how ZIR indexes have moved when a source file is updated.

The much better solution than doing this, would be to not store a new ZIR index into a Decl at all.

If you look further down this file, you can see this:

```zig
    if (lib_name != null) {
        // Look at the sema code for functions which has this logic, it just needs to
        // be extracted and shared by both var and func
        return sema.fail(block, src, "TODO: handle var with lib_name in Sema", .{});
    }
```

I suggest to take this comment's advice and solve that related problem while working on this problem.

And then finally, storing a extern lib name string into the `Module.Var` object in this function looks like a much more robust solution to the problem than messing with ZIR indexes or ZIR string table indexes. Be sure to copy the string data of `lib_name` so that it is not a reference to the ZIR memory, which gets freed with every update to the source file. You can copy it to the Decl's arena, or you can copy it with gpa, but then you will need to introduce a deinit() function for Module.Var objects.

As for functions, they should be handled pretty much the same thing as variables. Storing the lib name, if any, should be done on `Module.Fn` and `Module.Var`, not on `Module.Decl`.

Reply:
Done! Introduced new type called `Module.ExternFn` which contains a maybe-lib-name field that outlives ZIR and is tied to the liveness of `Module`. Similarly, I added a maybe-lib-name to `Module.Var` which the same liveness, and I've refactored out common  code path for analysing the parsed library name in Sema (for extern fns and vars). Have a look and let me know what you think!